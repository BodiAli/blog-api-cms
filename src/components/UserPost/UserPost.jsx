import { useEffect, useState } from "react";
import { useNavigate, useParams } from "react-router";
import parse from "html-react-parser";
import { toast } from "react-toastify";
import { format, formatDistanceToNow } from "date-fns";
import Loader from "../Loader/Loader";
import CommentCard from "../CommentCard/CommentCard";
import noImage from "../../assets/images/no-image.svg";
import styles from "./UserPost.module.css";

export default function UserPost() {
  const { postId } = useParams();
  const [post, setPost] = useState(null);
  const [loading, setLoading] = useState(true);
  const navigate = useNavigate();

  async function handleDeleteComment(comment) {
    try {
      const res = await fetch(
        `${import.meta.env.VITE_SERVER_URL}/posts/${comment.postId}/comments/${comment.id}`,
        {
          method: "DELETE",
          headers: {
            Authorization: localStorage.getItem("token"),
          },
        }
      );

      if (!res.ok) {
        throw new Error("Failed to delete comment please try again later");
      }

      toast.success("Comment deleted successfully!");
      setPost({
        ...post,
        Comments: post.Comments.filter((commentValue) => {
          return commentValue.id !== comment.id;
        }),
      });
    } catch (error) {
      toast.error(error.message);
    }
  }

  useEffect(() => {
    async function fetchUserPost() {
      try {
        const res = await fetch(`${import.meta.env.VITE_SERVER_URL}/posts/${postId}`);

        if (!res.ok) {
          if (res.status === 404) {
            const { error } = await res.json();
            throw new Error(error);
          }
          throw new Error("Failed to fetch post");
        }

        const post = await res.json();

        setPost(post);
      } catch (error) {
        toast.error(error.message);
        navigate("/", { viewTransition: true });
      } finally {
        setLoading(false);
      }
    }

    fetchUserPost();
  }, [postId, navigate]);

  if (loading) return <Loader />;

  return (
    <main className={styles.main}>
      <div className={styles.textContainer1}>
        <h2 className={styles.title}>{post?.title}</h2>
        <p className={styles.date}>
          <strong>Created at: </strong>
          {post && format(post.createdAt, "MMM d, y, H:m:s")}
        </p>
        <p className={styles.date}>
          <strong>Last updated: </strong>
          {post && formatDistanceToNow(post.updatedAt, { addSuffix: true, includeSeconds: true })}
        </p>
      </div>
      <div className={styles.textContainer2}>
        <p className={styles.likes}>
          <strong>Likes:</strong> {post?.likes}
        </p>
        <p className={styles.published}>{post?.published ? "Published" : "Unpublished"}</p>
      </div>
      <div className={styles.textContainer3}>
        <p>Topics:</p>
        <div>
          {post?.Topics.length === 0 ? (
            <span>This post has no topics</span>
          ) : (
            post?.Topics.map((topic, i) => {
              return <p key={i}>{topic.name}</p>;
            })
          )}
        </div>
      </div>
      <div className={styles.postContent}>
        <img src={post?.imgUrl ? post?.imgUrl : noImage} alt="Post cover" />
        <div className={styles.postContentHtml}>{post && parse(post.content)}</div>
      </div>

      <div className={styles.commentsContainer}>
        <div className={styles.headingContainer}>
          <h3>Comments</h3>
        </div>
        <div className={styles.commentsCardsContainer}>
          {post.Comments.length > 0 ? (
            post?.Comments.map((comment) => {
              return <CommentCard key={comment.id} comment={comment} onDelete={handleDeleteComment} />;
            })
          ) : (
            <p className={styles.noComments}>No comments yet!</p>
          )}
        </div>
      </div>
    </main>
  );
}
