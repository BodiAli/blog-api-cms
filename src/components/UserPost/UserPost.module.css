.main {
  padding: 1rem;
  font-family: "roboto", Arial, sans-serif;
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.textContainer1 {
  display: flex;
  align-items: center;
}

.textContainer1 > * {
  flex: 1;
}

.textContainer1 > h2 {
  font-size: 1.5rem;
}

.textContainer1 strong {
  font-weight: 500;
}

.textContainer1 > .date {
  text-align: center;
  color: #0000009c;
}

.date:last-child {
  text-align: end;
}

.textContainer2 {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 1rem;
}

.textContainer2 strong {
  font-weight: 500;
}

.textContainer2 > .published {
  background-color: rgb(72, 130, 180);
  color: white;
  padding: 0.4rem 0.8rem;
  border-radius: 5px;
}

.textContainer3 {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.textContainer3 > p {
  font-weight: 500;
}

.textContainer3 > div {
  display: flex;
  align-items: start;
  flex-wrap: wrap;
  gap: 1rem;
}

.textContainer3 > div > p {
  font-weight: 500;
  color: rgba(115, 113, 113, 0.761);
  background-color: rgba(128, 128, 128, 0.246);
  border-radius: 5px;
  padding: 0.3rem 0.6rem;
}

.postContent {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.postContent > img {
  width: 80%;
  border-radius: 5px;
  align-self: center;
}

.postContentHtml > * {
  all: revert;
}
