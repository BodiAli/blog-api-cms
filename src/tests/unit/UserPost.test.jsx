import { render, waitFor } from "@testing-library/react";
import { createRoutesStub } from "react-router";
import UserPost from "../../components/UserPost/UserPost";

const post = {
  id: 1,
  userId: 1,
  title: "Post 1",
  content: "Post content",
  published: true,
  likes: 5,
  createdAt: "2025-04-27T03:03:02.292Z",
  updatedAt: "2025-04-27T03:03:02.292Z",
  imgUrl: null,
  Topics: [
    { id: 1, name: "Topic 1" },
    { id: 2, name: "Topic 2" },
  ],
};

window.fetch = vi.fn(() => {
  return Promise.resolve({
    ok: true,
    status: 200,
    json: vi.fn(() => {
      return Promise.resolve(post);
    }),
  });
});

vi.mock("date-fns", async (importOriginal) => {
  const result = await importOriginal();
  return {
    ...result,
    formatDistanceToNow: vi.fn(() => "1 day ago"),
  };
});

const Stub = createRoutesStub([
  {
    path: "/users/posts/:postId",
    Component: () => <UserPost />,
  },
]);

describe("UserPost component", () => {
  test("Should render post data", async () => {
    let container;
    await waitFor(() => {
      const result = render(<Stub initialEntries={[`/users/posts/${post.id}`]} />);

      container = result.container;
    });

    expect(container).toMatchSnapshot();
  });
});
