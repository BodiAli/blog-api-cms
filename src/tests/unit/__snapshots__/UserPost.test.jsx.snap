// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserPost component > Should render post data 1`] = `
<div>
  <main
    class="_main_d6c027"
  >
    <div
      class="_textContainer1_d6c027"
    >
      <h2
        class="_title_d6c027"
      >
        Post 1
      </h2>
      <p
        class="_date_d6c027"
      >
        <strong>
          Created at: 
        </strong>
        2025/4/27, 6:3:2
      </p>
      <p
        class="_date_d6c027"
      >
        <strong>
          Last updated: 
        </strong>
        1 day ago
      </p>
    </div>
    <div
      class="_textContainer2_d6c027"
    >
      <p
        class="_likes_d6c027"
      >
        <strong>
          Likes:
        </strong>
         
        5
      </p>
      <p
        class="_published_d6c027"
      >
        Published
      </p>
    </div>
    <div
      class="_textContainer3_d6c027"
    >
      <p>
        Topics:
      </p>
      <div>
        <p>
          Topic 1
        </p>
        <p>
          Topic 2
        </p>
      </div>
    </div>
    <div
      class="_postContent_d6c027"
    >
      <img
        alt="Post cover"
        src="/src/assets/images/no-image.svg"
      />
      <div
        class="_postContentHtml_d6c027"
      >
        Post content
      </div>
    </div>
    <div
      class="_commentsContainer_d6c027"
    />
  </main>
</div>
`;
